File: backend-examples/go/preact_iso_url_pattern.go
// Run program: go run preact-iso-url-pattern.go

package main

import (
	// "fmt"
	"regexp"
	"strings"
)

func preactIsoUrlPatternMatch(urlStr, route string) bool {
	url := filterEmpty(strings.Split(urlStr, "/"))
	routeParts := filterEmpty(strings.Split(route, "/"))

	for i := 0; i < max(len(url), len(routeParts)); i++ {
		var m, param, flag string
		if i < len(routeParts) {
			re := regexp.MustCompile(`^(:?)(.*?)([+*?]?)$`)
			matches := re.FindStringSubmatch(routeParts[i])
			if len(matches) > 3 {
				m, param, flag = matches[1], matches[2], matches[3]
			}
		}

		var val string
		if i < len(url) {
			val = url[i]
		}

		// segment match:
		if m == "" && param == val {
			continue
		}

		// /foo/* match
		if m == "" && val != "" && flag == "*" {
			break
		}

		// segment mismatch / missing required field:
		if m == "" || (val == "" && flag != "?" && flag != "*") {
			return false
		}

		rest := flag == "+" || flag == "*"

		if rest {
			break
		}
	}

	return true
}

func filterEmpty(s []string) []string {
	var result []string
	for _, str := range s {
		if str != "" {
			result = append(result, str)
		}
	}
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Example usage:
// func main() {
// 	fmt.Println(preactIsoUrlPatternMatch("/foo/bar", "/foo/:param"))
// 	fmt.Println(preactIsoUrlPatternMatch("/foo/bar/baz", "/foo/*"))
// 	fmt.Println(preactIsoUrlPatternMatch("/foo", "/foo/:param?"))
// 	fmt.Println(preactIsoUrlPatternMatch("/foo/bar", "/bar/:param"))
// 	fmt.Println(preactIsoUrlPatternMatch("/users/test%40example.com/posts", "/users/:userId/posts"))
// }

File: backend-examples/python/preact_iso_url_pattern.py
# Run program: python3 preact-iso-url-pattern.py

def preact_iso_url_pattern_match(url, route):
    url = list(filter(None, url.split('/')))
    route = list(filter(None, (route or '').split('/')))

    for i in range(max(len(url), len(route))):
        m, param, flag = '', '', ''
        if i < len(route):
            parts = route[i].split(':')
            m = ':' if len(parts) > 1 else ''
            param = parts[-1]
            flag = ''
            if param and param[-1] in '+*?':
                flag = param[-1]
                param = param[:-1]

        val = url[i] if i < len(url) else None

        # segment match:
        if not m and param == val:
            continue

        # /foo/* match
        if not m and val and flag == '*':
            break

        # segment mismatch / missing required field:
        if not m or (not val and flag != '?' and flag != '*'):
            return False

        rest = flag in ('+', '*')

        if rest:
            break

    return True

# Example usage:
# print(preact_iso_url_pattern_match("/foo/bar", "/foo/:param"))
# print(preact_iso_url_pattern_match("/foo/bar/baz", "/foo/*"))
# print(preact_iso_url_pattern_match("/foo", "/foo/:param?"))
# print(preact_iso_url_pattern_match("/foo/bar", "/bar/:param"))
# print(preact_iso_url_pattern_match('/users/test%40example.com/posts', '/users/:userId/posts'))